Renews Maintenance Drones

JJRcop opened the pull request:
## About The Pull Request
It's the second try! (First: #48456)

Drones are little robots inspired by Keepers from the Mass Effect games. With laws to maintain both the station, and their distance from the matters of others, they were a great ghost role that let you fix up the station when you were otherwise out of the round. They were in the game for quite a while, but were abused. The abusive players completely ignored the laws and did whatever they wanted. Being tiny ventcrawlers, this became a huge issue for admins.

I attempted to [bring them back][48456], this time with some gameplay restrictions to enforce their laws a bit better. It seemed to go well, but I required headmin approval, and this all happened near a headmin election, so the project died.

Recently, some people have wanted to bring drones back. So, they gave me a ring, and I started a discussion with the admin team to see what other changes they would want in addition to the previous ones.

**Here are those changes:**
- They can't interact at all when close to another being, living or dead
- Their binary chat has been removed to separate them from AI and borgs
- They can only hold a whitelist of items necessary to do their job
- They can only interact with machines necessary to do their job
- The drone satchel is gone, and drones now have built in basic tools
- They also can't interact in some blacklisted high risk areas
- Finally, you must play 40 hours total as a Silicon role in order to play as a drone

Additionally, the drone dispenser is not mapped, and the drones must be researched to be constructed.

## Why It's Good For The Game

Drones are a fun way to pass the time, and this time they have a lot less freedom. You can still perform the basic function of maintaining the station, but not without difficulty and tedium. This will allow other station roles to perform much better than the drones, while still letting them chug along at their slower pace.


## Changelog
:cl: JJRcop
add: Maintenance Drones are once again constructable in Robotics.
del: The Drone Dispenser has not made a return.
config: New settings DRONE_REQUIRED_ROLE, and DRONE_ROLE_PLAYTIME, which determine the playtime exp required for playing drones.
/:cl:

[48456]: https://github.com/tgstation/tgstation/pull/48456

JJRcop committed:
Add shy component

Holders of the shy component can't use items on anyone but themselves
if there are living mobs around them that are not in a whitelist

JJRcop committed:
Drones are shy and pacifists

JJRcop committed:
Drones can no longer speak binary

JJRcop committed:
Update drone mob_spawn

JJRcop committed:
Add drones to techweb

JJRcop committed:
Shy component can't be shy from yourself

JJRcop committed:
Add missing params to some signal docs

JJRcop committed:
Shy component fixes
Blocks pull
Blocks UnarmedAttack (Because of the Pick up verb)

JJRcop committed:
Add shy room component
You can't interact when in a blacklisted room

JJRcop committed:
Shy component caches result for half a second

JJRcop committed:
Add shy room component to drone

JJRcop committed:
Add flavor text to drone spawner
And syndrone

JJRcop committed:
Add autodocs to shy and shy room

JJRcop committed:
Drones require 40 hours as a Silicon

JJRcop committed:
Drone time limit does not affect bold drones
(Drones without laws to avoid beings)

JJRcop committed:
Shy and Shy room block stripping

JJRcop committed:
Drones cannot use machines outside of a whitelist
These machines are for now:
`/obj/machinery/`
- `atmospherics`
- `autolathe`
- `cell_charger`
- `drone_dispenser`
- `light`
- `pipedispenser`
- `recharger`
- `rnd/production`

JJRcop committed:
Reformat drone's drone_machinery_whitelist_recursive

JJRcop committed:
Drones can only hold whitelisted items

JJRcop committed:
Add missing SIGNAL_HANDLER to shy, shy_room, technointrovert, and itempicky

JJRcop committed:
Add COMSIG_ITEM_PRE_UNEQUIP signal to doUnEquip, which can block it

JJRcop committed:
Replace Drone's satchel with built in tools

JJRcop committed:
Drones can pick up hats again

JJRcop committed:
Fix some Drone tool inhands

MrDoomBringer commented:
40 hours as silicon seems a little irrelevant. 
A gate lik e100 hours total is much more useful, both in terms of players knowing game etiquette and this being usable by people other than silicon mains (🤮)

IndieanaJones commented:
I'm noticing that drones don't seem to have any sort of internal metal storage.  Realistically, how much can they actually fix with the toolset provided here?

MrDoomBringer commented:
Should make drones be forced to use the system in #58029, ie they can only repair existing constructs

JJRcop commented:
> I'm noticing that drones don't seem to have any sort of internal metal storage. Realistically, how much can they actually fix with the toolset provided here?

Drones no longer have internal storage, but they still have hands. They can carry metal wherever they need it, but since they have no storage, they have to hold onto it, or drop it on the ground. This conveniently restricts what thet can hoard, as they can't hoard on themselves.

JJRcop commented:
> 40 hours as silicon seems a little irrelevant.
> A gate lik e100 hours total is much more useful, both in terms of players knowing game etiquette and this being usable by people other than silicon mains (🤮)

These numbers will not be changed. It is in a config file, so you are free to discuss with headmins after this is merged.

MrDoomBringer commented:
> These numbers will not be changed 😠 

Had a quick conversation with JJ about this, for anyone wondering on any reasoning there's a discord [policy discussion here ](https://discord.com/channels/326822144233439242/756312728414584934/822970047185551421)
The silicon hour requirement being linked with law-following players makes sense to me.

ATH1909 commented on code/modules/mob/living/simple_animal/friendly/drone/extra_drone_types.dm:
```diff
@@ -61,11 +62,15 @@
 	icon_state = "syndrone_item"
 	mob_name = "syndrone"
 	mob_type = /mob/living/simple_animal/drone/syndrone
+	short_desc = "You are a syndicate Maintenance Drone."
```
```suggestion
	short_desc = "You are a Syndicate Maintenance Drone."
```

ATH1909 commented on code/modules/research/techweb/all_nodes.dm:
```diff
@@ -304,6 +304,14 @@
+	id = "dex_robotics"
+	display_name = "Dexterous Robotics Research"
+	description = "The fine art of opposable thumbs."
+	prereq_ids = list("adv_engi", "adv_robotics", "biotech")
```
if you have to have posibrains researched before you can print drones, why are you printing drones instead of posibrains? just material concerns? in my experience, when robotics actually bothers to make posibrains, there are usually more posibrains than there are ghosts willing to take them- let alone ghosts with ~1.67 days' worth of silicon playtime.

CRITAWAKETS replied:
not only that but putting a posi that's marked to be an engiborg then putting it in an unsynced borg is just going to make a better drone than a drone

JJRcop replied:
This may be rebalanced later

ATH1909 replied:
> This may be rebalanced later, better to be underpowered than overpowered

is it, though? any PR that buffs drones is going to start another war in its comments section.

see: any PR that buffs PK borgs, or borgs in general

optimumtact replied:
it absoluteyl is

optimumtact replied:
and don't conflate people trying to make PK borgs into secborgs with changes to drones

ATH1909 commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -310,6 +332,30 @@
 		if(cleared)
 			to_chat(src, "--- [class] alarm in [A.name] has been cleared.")
 
+/mob/living/simple_animal/drone/proc/set_shy(newShy)
```
is hacking/emagging drones still a thing? if so, that should make them no longer shy, and factory resetting them should reset their shy variable to its default value for their (drone) type.

JJRcop replied:
Yes, hacking a drone properly sets shy to false, and factory resetting (using a wrench on a drone) will call the unhacking procedure which will set shy to `initial(shy)`

ATH1909 commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -121,6 +123,24 @@
+	var/list/drone_item_whitelist_recursive = list(	/obj/item/airlock_painter, /obj/item/conveyor_switch_construct, /obj/item/electronics, /obj/item/light, \
+													/obj/item/pipe_meter, /obj/item/stack/cable_coil, /obj/item/stack/circuit_stack, /obj/item/stack/conveyor, \
+													/obj/item/stack/pipe_cleaner_coil, /obj/item/stack/rods, /obj/item/stack/sheet, /obj/item/stack/tile, \
+													/obj/item/stock_parts, /obj/item/toner, /obj/item/clothing/head, /obj/item/clothing/mask )
```
will drones no longer be able to wear buckets (without somehow getting a human to help them, anyway)?

ATH1909 replied:
should rcds and rpds be added to this list? what about airlock electronics?

JJRcop replied:
Airlock electronics are an `/obj/item/electronics`

RCDs and RPDs intentionally missing: I don't want drones to use them. As I said in the PR body, I want them to be weaker.

I could add a bucket, if you deem it necessary to wear a bucket :wink:

ATH1909 replied:
well, the problem with adding a bucket is that a bucket can be used to hold reagents.

JJRcop replied:
Sorry bucket, you will be missed

ATH1909 commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -121,6 +123,24 @@
+															/obj/machinery/drone_dispenser, /obj/machinery/light, /obj/machinery/pipedispenser, /obj/machinery/recharger, \
+															/obj/machinery/rnd/production)
+	/// whitelisted drone items, direct
+	var/list/drone_item_whitelist_flat = list(	/obj/item/crowbar/drone, /obj/item/screwdriver/drone, /obj/item/wrench/drone, \
+												/obj/item/weldingtool/drone, /obj/item/wirecutters/drone )
```
why are multitools not in this list?

JJRcop replied:
I purposely left them out for now because I don't see them being necessary, and want to keep drones from being OP at first. I am open to adding them as well if they are acceptable, but probably not as a built in tool, just allow drones to use a normal multitool.

JJRcop replied:
I can see multitools necessary to fix telecomms, but maybe drones shouldn't be fixing telecomms.

CRITAWAKETS replied:
Its best to give them the multitool mostly just because derelict drones can and should build an ore silo

JJRcop replied:
Derilict drones can be modified separately

ATH1909 replied:
wait, are you making derelict drones shy as well, or will they remain as they were?

if the former, wow, this absolutely guts them.

JJRcop replied:
I haven't touched derelict drones with this PR, that's a good idea to test out

JJRcop committed:
Update syndrone spawn item description

Capitalize Syndicate

Co-authored-by: ATH1909 <42606352+ATH1909@users.noreply.github.com>

JJRcop commented:
I do not plan to make many changes aside from trivial ones, because this version of drones has headmin approval. Deviating from it is grounds to void the approval.

Mickyan commented:
Still, a bit disappointed about the hours requirement as drones seem like a perfect fit for easing new players into the silicon roles in a lower stress job, rather than the other way around

It feels like the time restrictions are a product of the well deserved stigma of old drones moreso than necessity, as this new version comes with a fair amount of limitations in comparison

Mickyan commented:
that's fair

Serxule commented:
40 hours as silicon is ludicrous considering no other ghost roles require time, including becoming a borg brain.

JJRcop commented:
> 40 hours as silicon is ludicrous considering no other ghost roles require time, including becoming a borg brain.

This is my last reply to this type of comment: This is not changing. Please do not comment about this.

JJRcop committed:
Drones can use right click

necromanceranne commented:
And thus begins the hunt for what shittery a drone can still get up too without hitting the blacklisted interactions.

JJRcop committed:
Fix Drone alarm runtimes

JJRcop committed:
Remove rustle from drone built in tools

ATH1909 commented on code/datums/components/technointrovert.dm:
```diff
@@ -0,0 +1,36 @@
+		message = _message
+
+/datum/component/technointrovert/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_CAN_USE_MACHINE, .proc/is_in_friend_group)
```
does this also prevent them from just bashing a machine or using one of their tools to interact with it?

can drones crowbar open doors, or will they have to ventcrawl around them?

can drones create machine frames or computer frames that they then can't interact with/deconstruct/bash to pieces?

JJRcop replied:
> does this also prevent them from just bashing a machine or using one of their tools to interact with it?

This just applies to using the machine, like with your hands

> can drones crowbar open doors, or will they have to ventcrawl around them?

Doors don't have the INTERACT_ATOM_ATTACK_HAND flag, so they ignore the can_interact proc completely

> can drones create machine frames or computer frames that they then can't interact with/deconstruct/bash to pieces?

machine frames and computer frames are not /obj/machinery, however I forgot computer circuitboards in the whitelist, thanks

ATH1909 commented on code/datums/components/itempicky.dm:
```diff
@@ -0,0 +1,35 @@
+	SIGNAL_HANDLER
+	if(!is_type_in_typecache(pickingup, whitelist))
+		to_chat(source, "<span class='warning'>[replacetext(message, "%TARGET", pickingup)]</span>")
+		return COMPONENT_CANT_USE_MACHINE
```
is COMPONENT_CANT_USE_MACHINE what we should be returning here?

JJRcop replied:
Good catch, no definitely not

ATH1909 commented on code/datums/components/itempicky.dm:
```diff
@@ -0,0 +1,35 @@
+/// You can only hold whitelisted items
```
can a human use the give command to give you an item, like what used to be done to bypass the xenomorph held item restrictions?

JJRcop replied:
I can't find a give command

JJRcop replied:
Give only attempts on carbon mobs, and drones are not carbons. Even if it did work, it uses the `put_in_hands` proc, so I forced that on the drone, and it fails when it's not on the whitelist. So this works correctly

JJRcop committed:
Fix itempicky returning COMPONENT_CANT_USE_MACHINE
Properly returns COMPONENT_LIVING_CANT_PUT_IN_HAND
These are both identical defines, but the correct one should still be used

LemonInTheDark commented on tgstation.dme:
```diff
@@ -540,6 +544,7 @@
 #include "code\datums\components\surgery_initiator.dm"
 #include "code\datums\components\swabbing.dm"
 #include "code\datums\components\swarming.dm"
+#include "code\datums\components\technointrovert.dm"
```
This include isn't sorted properly I don't think, are you doing this manually?

JJRcop replied:
yea. I will re-sort and be careful

Mothblocks commented on code/_onclick/other_mobs.dm:
```diff
@@ -1,3 +1,13 @@
+/// Checks for RIGHT_CLICK in modifiers and runs attack_hand_secondary if so. Returns TRUE if normal chain blocked
+/mob/living/proc/right_click_attack(atom/A, list/modifiers)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/_onclick/other_mobs.dm:
```diff
@@ -1,3 +1,13 @@
+/// Checks for RIGHT_CLICK in modifiers and runs attack_hand_secondary if so. Returns TRUE if normal chain blocked
+/mob/living/proc/right_click_attack(atom/A, list/modifiers)
+	if (LAZYACCESS(modifiers, RIGHT_CLICK))
```
Early return.

Mothblocks commented on code/_onclick/other_mobs.dm:
```diff
@@ -114,9 +119,15 @@
 */
+
+// A common early return for all UnarmedAttacks
+#define UNARMED_ATTACK_COMMON \
```
This is a crap macro, it's not obvious when reading any other code it's an early return, this is DRY gone wrong.

If you want anything to be a macro (and I'm not sure anything should be), make it the body of the if statement inside, and do the `if...return` everywhere else.

JJRcop replied:
I will reverse this and make these the start of the procs, no big deal

JJRcop replied:
How do I make this compliant with https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#no-duplicated-code, or are you asking me to ignore it this time?

Mothblocks replied:
Told you how to do it over Discord, but also in this same comment:

> make it the body of the if statement inside, and do the if...return everywhere else.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+	/// If we delete our parent when the holder gets deleted
+	var/del_parent_with_holder = FALSE
+
+/datum/component/holderloving/Initialize(_holder, _del_parent_with_holder)
```
Make these `holder` and `del_parent_with_holder` instead, and do `src.holder = holder` instead of `holder = _holder`.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+/datum/component/holderloving/RegisterWithParent()
+	RegisterSignal(holder, COMSIG_MOVABLE_MOVED, .proc/check_my_loc)
+	RegisterSignal(holder, COMSIG_PARENT_QDELETING, .proc/holder_deleting)
+	RegisterSignal(parent, list(	COMSIG_ITEM_DROPPED, COMSIG_ITEM_EQUIPPED, \
```
Do not indent midway through a line.

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#tabs-not-spaces

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+	RegisterSignal(holder, COMSIG_MOVABLE_MOVED, .proc/check_my_loc)
+	RegisterSignal(holder, COMSIG_PARENT_QDELETING, .proc/holder_deleting)
+	RegisterSignal(parent, list(	COMSIG_ITEM_DROPPED, COMSIG_ITEM_EQUIPPED, \
+									COMSIG_STORAGE_ENTERED, COMSIG_STORAGE_EXITED ), .proc/check_my_loc)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#tabs-not-spaces

Only one indentation level, and remove the space at the end.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+
+/datum/component/holderloving/UnregisterFromParent()
+	UnregisterSignal(holder, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	UnregisterSignal(	parent, list(	COMSIG_ITEM_DROPPED, COMSIG_ITEM_EQUIPPED, \
```
Fix indenting.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+/datum/component/holderloving/UnregisterFromParent()
+	UnregisterSignal(holder, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	UnregisterSignal(	parent, list(	COMSIG_ITEM_DROPPED, COMSIG_ITEM_EQUIPPED, \
+										COMSIG_STORAGE_ENTERED, COMSIG_STORAGE_EXITED ))
```
Fix indenting.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+/datum/component/holderloving
+	can_transfer = TRUE
+	/// Item that is kept track of
+	var/obj/item/holder
```
I'm confused. It looks like parent is an item, so then what is the holder if it's *also* an item?

JJRcop replied:
This component keeps the parent either directly inside holder, or inside of the holder's loc if that loc is a mob.

This is like defibs, the water tank backpack, etc. In this case the defib is the holder, and the paddles are the parent that the component is put in

JJRcop replied:
How would you suggest making this clearer?

Mothblocks replied:
Putting all this in the doc-comment, because I'm still confused 😄

Mothblocks replied:
I'm still confused and still think this needs to be explained in a doc comment.

JJRcop replied:
Does this help?

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,48 @@
+
+/datum/component/holderloving/proc/check_my_loc(datum/source)
+	SIGNAL_HANDLER
+	var/obj/item/Iparent = parent
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/itempicky.dm:
```diff
@@ -0,0 +1,35 @@
+	/// Message shown if you try to pick up an item not in the whitelist
+	var/message = "You don't like %TARGET, why would you hold it?"
+
+/datum/component/itempicky/Initialize(_whitelist, _message)
```
Same here--do the `src.whitelist = whitelist` trick.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+#define SHY_COMPONENT_CACHE_TIME 5 //deciseconds
```
Just do `0.5 SECONDS`, and you won't need this comment.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+	var/last_result = FALSE
+
+/// _shy_range, _whitelist, _message, dead_shy map to vars
+/datum/component/shy/Initialize(_whitelist, _shy_range, _message, _dead_shy)
```
Same here.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+		message = friend.message
+
+/// Returns TRUE or FALSE if you are within shy_range tiles from a /mob/living
+/datum/component/shy/proc/bashful(atom/A)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks replied:
Also, this should probably be `is_bashful` to make it obvious it returns a boolean.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+
+
+
+/datum/component/shy/proc/bashful_click(datum/source, atom/A, params)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+				break
+
+	last_result = result
+	result_expires = world.time + SHY_COMPONENT_CACHE_TIME
```
Use cooldown defines instead.

JJRcop replied:
Which file contains these?

Mothblocks replied:
Not sure, but search COOLDOWN_DECLARE.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+	SIGNAL_HANDLER
+	return bashful(A) && COMSIG_MOB_CANCEL_CLICKON
+
+/datum/component/shy/proc/bashful_pull(datum/source, atom/movable/AM, force)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,93 @@
+/datum/component/shy/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOB_CLICKON, .proc/bashful_click)
+	RegisterSignal(parent, COMSIG_LIVING_TRY_PULL, .proc/bashful_pull)
+	RegisterSignal(parent, list(COMSIG_LIVING_UNARMED_ATTACK, COMSIG_HUMAN_EARLY_UNARMED_ATTACK), .proc/bashful_unarmed)
```
Bad names, make these all `on_click`, `on_try_pull`, etc.

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+	var/message = "%ROOM is too creepy to do that!"
+
+/// _blacklist, and _message map to vars
+/datum/component/shy_room/Initialize(_blacklist, _message)
```
Same here.

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+/// You can't use items on anyone other than yourself if you stand in a blacklisted room
+/datum/component/shy_room
```
Rename it `shy_in_room` or similar, this implies the room itself is shy.

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+
+/datum/component/shy_room/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOB_CLICKON, .proc/frightened_click)
+	RegisterSignal(parent, COMSIG_LIVING_TRY_PULL, .proc/frightened_pull)
```
Again, make all these `on_xxx`.

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+		message = friend.message
+
+/// Returns TRUE or FALSE if you are in a blacklisted area
+/datum/component/shy_room/proc/frightened(atom/A)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+		to_chat(owner, "<span class='warning'>[replacetext(message, "%ROOM", room)]</span>")
+		return TRUE
+
+/datum/component/shy_room/proc/frightened_click(datum/source, atom/A, params)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+	SIGNAL_HANDLER
+	return frightened(A) && COMSIG_MOB_CANCEL_CLICKON
+
+/datum/component/shy_room/proc/frightened_pull(datum/source, atom/movable/AM, force)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+	SIGNAL_HANDLER
+	return frightened(AM) && COMSIG_LIVING_CANCEL_PULL
+
+/datum/component/shy_room/proc/frightened_unarmed(datum/source, atom/target, proximity, modifiers)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/shy_room.dm:
```diff
@@ -0,0 +1,62 @@
+	SIGNAL_HANDLER
+	return frightened(target) && COMPONENT_CANCEL_ATTACK_CHAIN
+
+/datum/component/shy_room/proc/frightened_nightclub(datum/source, atom/target, obj/item/equipping)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/datums/components/technointrovert.dm:
```diff
@@ -0,0 +1,36 @@
+	/// Message presented when
+	var/message = "That %TARGET is strange! Let's avoid it."
+
+/datum/component/technointrovert/Initialize(_whitelist, _message)
```
Same here.

Mothblocks commented on code/datums/elements/strippable.dm:
```diff
@@ -74,6 +74,8 @@
 /// This should be used for checking if an item CAN be equipped.
 /// It should not perform the equipping itself.
 /datum/strippable_item/proc/try_equip(atom/source, obj/item/equipping, mob/user)
+	if(SEND_SIGNAL(user, COMSIG_CAN_STRIP, source, user, equipping) & COMPONENT_CANT_STRIP)
```
This is a bad signal. Signals should say what's going on, not be a request. Change it to `COMSIG_TRY_STRIP` or similar, just like everything else.

Mothblocks commented on code/game/machinery/_machinery.dm:
```diff
@@ -342,6 +342,9 @@
 	if(!isliving(user))
 		return FALSE //no ghosts in the machine allowed, sorry
 
+	if(SEND_SIGNAL(user, COMSIG_CAN_USE_MACHINE, src) & COMPONENT_CANT_USE_MACHINE)
```
Same here. Signals are not a request, they are an announcement.

Mothblocks commented on code/modules/mob/inventory.dm:
```diff
@@ -186,7 +186,8 @@
 
 /mob/living/put_in_hand_check(obj/item/I)
-	if(istype(I) && ((mobility_flags & MOBILITY_PICKUP) || (I.item_flags & ABSTRACT)))
+	if(istype(I) && ((mobility_flags & MOBILITY_PICKUP) || (I.item_flags & ABSTRACT)) && \
```
Change the new line to be before the `&&` so we see the next line is `&&`.

Mothblocks commented on code/modules/mob/living/carbon/carbon_stripping.dm:
```diff
@@ -10,7 +10,8 @@
 
 /datum/strippable_item/mob_item_slot/back/alternate_action(atom/source, mob/user)
-	return strippable_alternate_action_internals(get_item(source), source, user)
+	if(!..())
```
Why is this one `!..()` while the others are `..()`?

JJRcop replied:
This is a failure on my part to document on the parent. It currently returns `TRUE` when the alternate action is blocked. It should be changed to return `TRUE` by default and `FALSE` if blocked, which makes more sense.
https://github.com/tgstation/tgstation/blob/be4cd777eefcc472e8b52688e60f78fe0b4cf6c0/code/datums/elements/strippable.dm#L184-L187

Mothblocks commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -108,7 +112,8 @@ GLOBAL_LIST_INIT(strippable_human_items, create_strippable_list(list(
 
 /datum/strippable_item/mob_item_slot/suit_storage/alternate_action(atom/source, mob/user)
-	return strippable_alternate_action_internals(get_item(source), source, user)
+	if (!..())
```
Same here, I don't understand the consistency.

Mothblocks commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -122,7 +127,8 @@ GLOBAL_LIST_INIT(strippable_human_items, create_strippable_list(list(
 
 /datum/strippable_item/mob_item_slot/belt/alternate_action(atom/source, mob/user)
-	return strippable_alternate_action_internals(get_item(source), source, user)
+	if (!..())
```
Same here.

Mothblocks commented on code/datums/elements/strippable.dm:
```diff
@@ -178,6 +182,9 @@
 /// `has_alternate_action` needs to be TRUE.
 /datum/strippable_item/proc/alternate_action(atom/source, mob/user)
+	SHOULD_CALL_PARENT(TRUE)
+	if(SEND_SIGNAL(user, COMSIG_CAN_STRIP, source) & COMPONENT_CANT_STRIP)
```
This doesn't make sense to be on the same signal (even when you do make it `TRY_STRIP`).

Signals, generally, should come from *one* place. Furthermore, this doesn't even have the same signature, since it doesn't provide the item. Make a separate signal that someone is trying to use an alternate action.

JJRcop replied:
I don't know what I was smoking

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -121,6 +123,24 @@
+	/// whitelisted drone machines, direct
+	var/list/drone_machinery_whitelist_flat
+	/// whitelisted drone machines, recursive/includes descendants
+	var/list/drone_machinery_whitelist_recursive = list(	/obj/machinery/atmospherics, /obj/machinery/autolathe, /obj/machinery/cell_charger, /obj/machinery/disposal, \
+															/obj/machinery/drone_dispenser, /obj/machinery/light, /obj/machinery/pipedispenser, /obj/machinery/recharger, \
+															/obj/machinery/rnd/production)
+	/// whitelisted drone items, direct
+	var/list/drone_item_whitelist_flat = list(	/obj/item/crowbar/drone, /obj/item/screwdriver/drone, /obj/item/wrench/drone, \
+												/obj/item/weldingtool/drone, /obj/item/wirecutters/drone )
```
Fix all this hideous indenting.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -121,6 +123,24 @@
+	var/list/drone_item_whitelist_flat = list(	/obj/item/crowbar/drone, /obj/item/screwdriver/drone, /obj/item/wrench/drone, \
+												/obj/item/weldingtool/drone, /obj/item/wirecutters/drone )
+	/// whitelisted drone items, recursive/includes descendants
+	var/list/drone_item_whitelist_recursive = list(	/obj/item/airlock_painter, /obj/item/conveyor_switch_construct, /obj/item/electronics, /obj/item/light, \
+													/obj/item/pipe_meter, /obj/item/stack/cable_coil, /obj/item/stack/circuit_stack, /obj/item/stack/conveyor, \
+													/obj/item/stack/pipe_cleaner_coil, /obj/item/stack/rods, /obj/item/stack/sheet, /obj/item/stack/tile, \
+													/obj/item/stock_parts, /obj/item/toner, /obj/item/clothing/head, /obj/item/clothing/mask )
```
More hideous indenting.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -310,6 +332,30 @@
+		LoadComponent(/datum/component/itempicky, drone_good_items, "Using %TARGET could break your laws.")
+	else
+		REMOVE_TRAIT(src, TRAIT_PACIFISM, INNATE_TRAIT)
+		var/list/qdel_components = list(	GetComponent(/datum/component/shy), GetComponent(/datum/component/shy_room), \
+											GetComponent(/datum/component/technointrovert), GetComponent(/datum/component/itempicky))
```
Fix the indenting.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -310,6 +332,30 @@
+	var/list/drone_good_machines = make_associative(drone_machinery_whitelist_flat) + typecacheof(drone_machinery_whitelist_recursive)
+	var/list/drone_good_items = make_associative(drone_item_whitelist_flat) + typecacheof(drone_item_whitelist_recursive)
+	if(shy)
+		ADD_TRAIT(src, TRAIT_PACIFISM, INNATE_TRAIT)
```
This isn't an innate trait if it's conditional, make a new source.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -310,6 +332,30 @@
 		if(cleared)
 			to_chat(src, "--- [class] alarm in [A.name] has been cleared.")
 
+/mob/living/simple_animal/drone/proc/set_shy(newShy)
```
`new_shy`

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+	builtintools += new /obj/item/weldingtool/drone(src)
+	builtintools += new /obj/item/wirecutters/drone(src)
+
+	for(var/_tool in builtintools)
+		var/obj/item/tool = _tool
```
Use `as anything`.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+	. = ..()
+	var/static/list/drone_builtins = list(	/obj/item/crowbar/drone, /obj/item/screwdriver/drone, /obj/item/wrench/drone, \
+											/obj/item/weldingtool/drone, /obj/item/wirecutters/drone)
+	var/datum/component/storage/STR = GetComponent(/datum/component/storage)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+
+/obj/item/storage/drone_tools/ComponentInitialize()
+	. = ..()
+	var/static/list/drone_builtins = list(	/obj/item/crowbar/drone, /obj/item/screwdriver/drone, /obj/item/wrench/drone, \
+											/obj/item/weldingtool/drone, /obj/item/wirecutters/drone)
```
Fix indenting.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+
+/obj/item/storage/drone_tools
+	name = "built-in tools"
+	desc = "Access your built in tools"
```
```suggestion
	desc = "Access your built-in tools."
```

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+
```
Random new line.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+/obj/item/screwdriver/drone/worn_overlays(isinhands = FALSE, icon_file)
+	. = list()
+	if(isinhands)
+		var/mutable_appearance/M = mutable_appearance(icon_file, "screwdriver_head")
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

JJRcop replied:
I think this is a bit too far but ok

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drone_tools.dm:
```diff
@@ -0,0 +1,84 @@
+		tool.AddComponent(/datum/component/holderloving, src, TRUE)
+
+
+///// DRONE TOOLS //////
```
Don't do this, it messes with the autodoc.

Furthermore, it should be obvious these are the drone tools, they literally say drone and are in the drone tools file.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -30,12 +34,18 @@
-			return
-	. = ..()
+/obj/effect/mob_spawn/drone/allow_spawn(mob/user)
+	var/client/C = user.client
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -30,12 +34,18 @@
+/obj/effect/mob_spawn/drone/allow_spawn(mob/user)
+	var/client/C = user.client
+	var/mob/living/simple_animal/drone/drone_type = mob_type
+	if(initial(drone_type.shy) && C && CONFIG_GET(flag/use_exp_restrictions_other))
```
Early return all these.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -30,12 +34,18 @@
+		var/required_playtime = CONFIG_GET(number/drone_role_playtime) * 60
+		if(required_playtime <= 0)
+			return ..()
+		var/current_playtime = C.calc_exp_type(required_role)
```
From what I've been told, clients can disappear at any time. Just do `?.`.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -30,12 +34,18 @@
+		var/current_playtime = C.calc_exp_type(required_role)
+		if (current_playtime < required_playtime)
+			var/minutes_left = required_playtime - current_playtime
+			var/playtime_left = DisplayTimeText(minutes_left MINUTES)
```
Don't do `MINUTES` like this, it's ugly. Just do `* (1 MINUTES)`.

Mothblocks commented on tgstation.dme:
```diff
@@ -2638,6 +2643,7 @@
 #include "code\modules\mob\living\simple_animal\friendly\sloth.dm"
 #include "code\modules\mob\living\simple_animal\friendly\snake.dm"
 #include "code\modules\mob\living\simple_animal\friendly\drone\_drone.dm"
+#include "code\modules\mob\living\simple_animal\friendly\drone\drone_tools.dm"
```
This is out of order too.

Mothblocks reviewed (changes requested):
I make no comment on whether or not I want this feature or not, I defer to Lemon since the design space this intrudes upon is mostly his.

JJRcop committed:
Drones can hold circuitboards for machines and computers

JJRcop committed:
Alphabetize tgstation.dme errors

JJRcop committed:
Correct indentation mistakes with drones

JJRcop committed:
Rename code/datums/components/shy_room.dm -> code/datums/components/shy_in_room.dm

JJRcop committed:
Change _arg to arg, in holderloving, itempicky, shy, and shy_in_room

JJRcop committed:
Move AND to next line in multi-line IF

Timberpoes commented:
I don't see the point of adding these to robotics.

They're completely non-interactive. You can't talk to them, you can't interact with them in any meaningful way, you can't change their laws, they just exist to be sentient NPCs. This makes them less interesting for the shift as a whole than roboticists just making more cyborgs.

I can see absolutely no reason for these drones to exist when cyborgs already exist and when cyborgs contribute far, far more to the shift in terms of player-to-player interaction than mute, impotent drones.

Your justification appears to be "they're a tedious ghost role that isn't allowed to interact with anything and does everything badly" - I'm really not seeing it.

JJRcop commented:
They have an excellent advantage of ventcrawling, which is a huge increase to mobility. While they are inferior to cyborgs, they are certainly easier to make, and cost less materials for each one. This is nice in a pinch if you need them now.

I look at them as a light weight alternative that can do a lot with little.

JJRcop committed:
Use time define for SHY_COMPONENT_CACHE_TIME

optimumtact commented:
A lot of people when dead, wish to occupy themselves with something to pass the time but no longer wish to actually involve themself in the meta of the round.

Drones fufil that sort of role and were popular among a proportion of the players, the removal PR was contentious for that reason.

Timberpoes commented:
I see. In that context I can see why they'd be nice as a ghost role. They still seem fairly pointless to me, but I will happily cede to the judgment of my peers who have first hand experience with old drones and the reasons why they existed to begin with.

JJRcop committed:
Shy uses cooldown defines for cache

JJRcop committed:
Rename signals COMSIG_CAN_(STRIP | USE_MACHINE) -> COMSIG_TRY_(STRIP | USE_MACHINE)

JJRcop committed:
Add COMSIG_TRY_ALT_ACTION

JJRcop committed:
Rename procs to be more self-evident in holderloving, shy, shy_in_room, and technointrovert

JJRcop committed:
Add DRONE_SHY_TRAIT trait source, use it instead of INNATE for pacifism

JJRcop committed:
Rename arg in drone/set_shy

necromanceranne commented:
I wouldn't mind a few roundstart drone bodies just kept inside the AI sat or some similar location before the printing machine is researched, but I can get why that might also be an unpopular choice if we want to keep the decision to include drones in a round to the players who currently occupy that round.

JJRcop committed:
Cleanup _onclick/other_mobs.dm blunders
rename proc/right_click_attack -> proc/right_click_attack_chain
UNARMED_ATTACK_COMMON macro replaced with early return of modified macro

JJRcop committed:
drone_tools formatting and var name fixes

JJRcop committed:
drones_as_items.dm cleanup and formatting fixes

JJRcop committed:
Flip return value of /datum/strippable_item/proc/alternate_action

JJRcop committed:
Remove _ from args in technointrovert

JJRcop committed:
Fix drone spawn corpse

JJRcop committed:
Rename screwdriver head Mutable Appearance

JJRcop committed:
Autodoc holderloving with a better description

JJRcop committed:
Add classic drone, and make N.E.R.D use classic drone

JJRcop commented:
Added a classic drone subtype for admins, with the old satchel and most of the restrictions disabled

JJRcop commented:
Another concern: Right now the supermatter area is focused on the little chamber housing the crystal itself, so drones can mess with the lasers and plumbing around the crystal, which is clnsidered Engineering rather than the Supermatter.

I have two options:
 - Remap every station to extend the Supermatter area far enough to include all of the ducting and lasers
 - Blacklist Engineering completely.

I would prefer the first option, but right now I worry about mapping conflicts. However, the supermatter crystal isn't touched much. Which choice is better?

Dorsisdwarf commented:
Personally I think that rather than extending the SM area you ought to create a new "Supermatter Engine Room" area and use that, since the current expectation is that the "Supermatter" area is just the chamber itself but the "Engineering Main" is wayyyy too large an area

JJRcop commented:
> Personally I think that rather than extending the SM area you ought to create a new "Supermatter Engine Room" area and use that, since the current expectation is that the "Supermatter" area is just the chamber itself but the "Engineering Main" is wayyyy too large an area

I like this solution.

JJRcop committed:
Add Supermatter Engine Room area and mapped to infrastructure surrounding engine
Drones can't interact with this new area either

JJRcop commented:
@Mothblocks I believe I have addressed all of your concerns

Mothblocks commented on code/_onclick/other_mobs.dm:
```diff
@@ -114,8 +120,13 @@
+
+// If the UnarmedAttack chain is blocked
+#define LIVING_UNARMED_ATTACK_BLOCKED (HAS_TRAIT(src, TRAIT_HANDS_BLOCKED) \
+		|| SEND_SIGNAL(src, COMSIG_LIVING_UNARMED_ATTACK, A, proximity_flag, modifiers) & COMPONENT_CANCEL_ATTACK_CHAIN)
```
One indent.

Mothblocks replied:
Also, make this macro take `A` as an argument, so that we can change the variable names later without much problem (it's also just bad form to act unhygienically).

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,69 @@
+	if(!isitem(parent) || !holder)
+		return COMPONENT_INCOMPATIBLE
+	src.holder = holder
+	src.del_parent_with_holder = (del_parent_with_holder || FALSE)
```
```suggestion
	src.del_parent_with_holder = del_parent_with_holder
```

You're not checking explicitly for `== FALSE` later anyway, so all your checks will work fine.

JJRcop replied:
I did this because I just don't want del_parent_with_holder to be null

Mothblocks replied:
It doesn't matter, but if you want that then do `if (del_parent_with_holder) src.del_parent_with_holder = ...`, for consistency.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,69 @@
+	), .proc/check_my_loc)
+
+/datum/component/holderloving/UnregisterFromParent()
+	UnregisterSignal(holder, list(COMSIG_MOVABLE_MOVED, COMSIG_PARENT_QDELETING))
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_DROPPED,
+		COMSIG_ITEM_EQUIPPED,
+		COMSIG_STORAGE_ENTERED,
+		COMSIG_STORAGE_EXITED,
+	))
```
Unregister in one list.

JJRcop replied:
holder and parent are different atoms, I don't see what you mean. Do you mean remove the indentation?

Mothblocks replied:
Oh, nevermind.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,103 @@
+		message = friend.message
+
+/// Returns TRUE or FALSE if you are within shy_range tiles from a /mob/living
+/datum/component/shy/proc/is_shy(atom/A)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,103 @@
+
+
+
+/datum/component/shy/proc/on_clickon(datum/source, atom/A, params)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/datums/components/shy.dm:
```diff
@@ -0,0 +1,103 @@
+	SIGNAL_HANDLER
+	return is_shy(A) && COMSIG_MOB_CANCEL_CLICKON
+
+/datum/component/shy/proc/on_try_pull(datum/source, atom/movable/AM, force)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/datums/components/shy_in_room.dm:
```diff
@@ -0,0 +1,74 @@
+		message = friend.message
+
+/// Returns TRUE or FALSE if you are in a blacklisted area
+/datum/component/shy_in_room/proc/is_shy(atom/A)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/datums/components/shy_in_room.dm:
```diff
@@ -0,0 +1,74 @@
+		to_chat(owner, "<span class='warning'>[replacetext(message, "%ROOM", room)]</span>")
+		return TRUE
+
+/datum/component/shy_in_room/proc/on_clickon(datum/source, atom/A, params)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/datums/components/shy_in_room.dm:
```diff
@@ -0,0 +1,74 @@
+	SIGNAL_HANDLER
+	return is_shy(A) && COMSIG_MOB_CANCEL_CLICKON
+
+/datum/component/shy_in_room/proc/on_try_pull(datum/source, atom/movable/AM, force)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Single letter variable names are not allowed. Name this what it is. Read the linked guide for more information on *how* to do this.

Mothblocks commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -108,7 +112,8 @@ GLOBAL_LIST_INIT(strippable_human_items, create_strippable_list(list(
 
 /datum/strippable_item/mob_item_slot/suit_storage/alternate_action(atom/source, mob/user)
-	return strippable_alternate_action_internals(get_item(source), source, user)
+	if (..())
```
Make all of these `if (!..()) return`, like your other ones.

Mothblocks commented on code/modules/mob/living/carbon/human/human_stripping.dm:
```diff
@@ -122,7 +127,8 @@ GLOBAL_LIST_INIT(strippable_human_items, create_strippable_list(list(
 
 /datum/strippable_item/mob_item_slot/belt/alternate_action(atom/source, mob/user)
-	return strippable_alternate_action_internals(get_item(source), source, user)
+	if (..())
```
`if (!..()) return`

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:
```diff
@@ -310,6 +360,34 @@
+		LoadComponent(/datum/component/itempicky, drone_good_items, "Using %TARGET could break your laws.")
+	else
+		REMOVE_TRAIT(src, TRAIT_PACIFISM, DRONE_SHY_TRAIT)
+		var/list/qdel_components = list(
+			GetComponent(/datum/component/shy),
+			GetComponent(/datum/component/shy_in_room),
+			GetComponent(/datum/component/technointrovert),
+			GetComponent(/datum/component/itempicky),
+		)
+		for(var/component in qdel_components)
+			qdel(component)
```
You're not saving any lines or readability with this list and loop. Just use `RemoveComponent`, which I'm pretty sure exists.

JJRcop replied:
`RemoveComponent` is a member of components. This is the way you must do it. I can use `qdel(GetComponent(/datum/component/dead_man_walking))` if you'd like

Mothblocks replied:
Yes, just qdel.

Mothblocks commented on code/modules/mob/living/simple_animal/friendly/drone/drones_as_items.dm:
```diff
@@ -30,12 +34,19 @@
+/obj/effect/mob_spawn/drone/allow_spawn(mob/user)
+	var/client/user_client = user.client
+	var/mob/living/simple_animal/drone/drone_type = mob_type
+	if(!(initial(drone_type.shy) && user_client && CONFIG_GET(flag/use_exp_restrictions_other)))
```
```suggestion
	if(!initial(drone_type.shy) || isnull(user_client) || !CONFIG_GET(flag/use_exp_restrictions_other))
```

Mothblocks replied:
Easier to read.

JJRcop committed:
Add readability suggestion to drone/allow_spawn

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

JJRcop committed:
_onclick/other_mobs.dm argument name corrections

JJRcop committed:
More descriptive autodoc for holderloving/holder

JJRcop committed:
holderloving/Initialize cleanup

JJRcop committed:
Shy and shy_in_room argument specificity update

JJRcop committed:
strippable_item alternate_action early return consistency

JJRcop committed:
Replace component loop in drone shy_update with static qdels

JJRcop committed:
holderloving var/obj/item/holder -> var/atom/holder
allows more flexibility

JJRcop committed:
strippable_item alternate_action early return consistency 2 of 2(?)

JJRcop committed:
Fix extra arg to COMSIG_TRY_STRIP in strippable_item/try_equip

JJRcop commented:
Remaining requested changes resolved.

Mothblocks commented on code/datums/components/holderloving.dm:
```diff
@@ -0,0 +1,73 @@
+ */
+/datum/component/holderloving
+	can_transfer = TRUE
+	/** Item that parent is bound to.
+	 * We try to keep parent either directly in holder, or in holder's loc if loc is a mob,
```
```suggestion
	/**
	 * Item that parent is bound to.
	 * We try to keep parent either directly in holder, or in holder's loc if loc is a mob,
```

I believe this is required for doc comments.

JJRcop replied:
Looks better too, in either case

Mothblocks reviewed (approved):
One minor nitpick, but the *code* looks serviceable to me. I still would like to hear more opinions about the design from maintainers, specifically @LemonInTheDark

JJRcop commented:
What are the next steps for this patch?

optimumtact commented on code/__DEFINES/dcs/signals.dm:
```diff
@@ -290,8 +290,15 @@
 #define COMSIG_LIVING_GET_PULLED "living_start_pulled"
+///called on /living, when pull is attempted, but before it completes, from base of [/mob/living/start_pulling]: (atom/movable/thing, force)
+#define COMSIG_LIVING_TRY_PULL "living_try_pull"
+	#define COMSIG_LIVING_CANCEL_PULL (1 << 0)
```
why are you doing this

JJRcop replied:
Players can right click mobs to get the pull verb, which dodged the components I set up to block clicking.

JJRcop replied:
I welcome suggestions to do this better, maybe making the pull verb go down the code path of ctrl click instead

JJRcop replied:
Further discussed in discord: issue was about indenting a define, not for having a pull signal
Answer: following precedent
Conclusion: precedent sucks

Happy to fix the whole file for you

LemonInTheDark commented:
> 
> 
> One minor nitpick, but the _code_ looks serviceable to me. I still would like to hear more opinions about the design from maintainers, specifically @LemonInTheDark

For the record I was initially concerned about drones because I was worried they'd outpace atmos techs and engineers, but in a not totally roundstart form like this I've got no problems with em

Farquaar commented:
I like the PR but I don't get why so many comments are being marked as spam when they're clearly discussing the PR.

theta006 commented:
40 hours is a lot just to play as a bad engi borg

optimumtact commented:
so go play engi borg then.

These are explicitly for dead players to fuck about in until teh round ends when they no longer want to rejoin the round meta

davethwave commented:
Now to include a suicide with a drone shell inhand to boot your soul into the drone the cool way.

notfrying1pans commented:
>so go play engi borg then.

except that the average person who would want to play engie borg probably doesnt have over 40 hours as a borg

Mothblocks commented:
if you think the 40 hour minimum is too high take it up with head admins over on the policy discussion board
